realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -66,6 +69,8 @@
     RealmSchema schema;
     Handler handler;
     HandlerController handlerController;
+    List<Runnable> asyncTransactionCallbacks = new ArrayList<Runnable>();
+    List<Future<?>> asyncTransactions = new ArrayList<Future<?>>();
```

Comment:
`ArrayList` is not thread safe, but it is accessed in caller thread and the async transaction thread.